rest/client/iter.go

```diff
@@ -0,0 +1,54 @@
+package client
+
+// Iter defines an iterator type that should be used when implementing list methods. It's intended to
+// be embedded in a domain specific iterator struct.
+type Iter struct {
+	query Query
+
+	page    ListResponse
+	current interface{}
+	results []interface{}
+
+	err error
+}
+
+// Query defines a closure that domain specific iterators must implement. The implementation should
+// include a call to the API and should return the API response with a separate slice of the results.
+type Query func(string) (ListResponse, []interface{}, error)
+
+// GetIter returns a new initialized iterator. This method automatically makes the first query to
+// populate the results. List methods should use this helper method when building domain specific
+// iterators.
+func GetIter(url string, query Query) *Iter {
+	it := &Iter{query: query}
+	it.page, it.results, it.err = it.query(url)
+	return it
+}
+
+// Next moves the iterator to the next result.
+func (it *Iter) Next() bool {
+	if len(it.results) == 0 && it.page.NextPageURL() != "" {
+		it.page, it.results, it.err = it.query(it.page.NextPageURL())
+	}
+	if it.err != nil || len(it.results) == 0 {
+		return false
+	}
+	it.current = it.results[0]
+	it.results = it.results[1:]
+	return true
+}
```

Comment:
These aren't the same semantics we use for our own internal clients so I would like more edge testing to gain confidence.

Reply:
makes sense! i based this partially on our own pattern and partially on [stripe's](https://github.com/stripe/stripe-go/blob/master/iter.go#L62). i was planning to add some tests for this iter type itself so i'll go ahead and do that in this PR

Reply:
just added some tests for the iterator pattern

Reply:
@calebcase makes sense, i'll add that